samples/go/photo-date-dedup/main.go

```diff
@@ -0,0 +1,186 @@
+	fmt.Fprintf(os.Stderr, "Usage: %s [flags] [input-paths...]\n\n", path.Base(os.Args[0]))
+	fmt.Fprintln(os.Stderr, "Flags:\n")
+	flag.PrintDefaults()
+}
+
+func index() (win bool) {
+	var dbStr = flag.String("db", "", "input database spec")
+	var outDSStr = flag.String("out-ds", "", "output dataset to write to - if empty, defaults to input dataset")
+	var thresh = flag.Int("threshold", 5000, "Number of milliseconds within which to consider photos duplicates")
+	verbose.RegisterVerboseFlags(flag.CommandLine)
+
+	flag.Usage = usage
+	flag.Parse(false)
+
+	if flag.NArg() == 0 {
+		flag.Usage()
+		return
+	}
+
+	cfg := config.NewResolver()
+	db, err := cfg.GetDatabase(*dbStr)
+	if err != nil {
+		fmt.Fprintf(os.Stderr, "Invalid input database '%s': %s\n", flag.Arg(0), err)
+		return
+	}
+	defer db.Close()
+
+	var outDS datas.Dataset
+	if !datas.IsValidDatasetName(*outDSStr) {
+		fmt.Fprintf(os.Stderr, "Invalid output dataset name: %s\n", *outDSStr)
+		return
+	} else {
+		outDS = db.GetDataset(*outDSStr)
+	}
+
+	inputs, err := spec.ReadAbsolutePaths(db, flag.Args()...)
+	if err != nil {
+		fmt.Fprintf(os.Stderr, "%s\n", err)
+		return
+	}
+
+	byDate := buildDateIndex(db, inputs)
+	groups := buildGroups(db, *thresh, byDate)
+
+	val, err := marshal.Marshal(struct {
+		ByDate types.Map
+		Groups types.List
+	}{
+		byDate,
+		groups,
+	})
+	d.Chk.NoError(err)
+
+	meta, err := marshal.Marshal(struct {
+		Date string
+	}{
+		time.Now().Format(spec.CommitMetaDateFormat),
+	})
+	d.Chk.NoError(err)
+
+	outDS, err = db.Commit(outDS, val, datas.CommitOptions{
+		Meta: meta.(types.Struct),
+	})
+	if err != nil {
+		fmt.Fprintf(os.Stderr, "Could not commit: %s\n", err)
+		return
+	}
+	return
+}
+
+func buildDateIndex(db types.ValueReadWriter, inputs []types.Value) types.Map {
+	indexBuilder := types.NewGraphBuilder(db, types.MapKind, true)
+	for _, v := range inputs {
+		walk.WalkValues(v, db, func(cv types.Value) (stop bool) {
+			var p Photo
+			if err := marshal.Unmarshal(cv, &p); err == nil {
+				stop = true
+				indexBuilder.SetInsert(
+					[]types.Value{types.Number(float64(p.DateTaken.NsSinceEpoch))},
+					cv)
+			}
+			return
+		})
+	}
+	return indexBuilder.Build().(types.Map)
+}
+
+func buildGroups(db types.ValueReadWriter, thresh int, byDate types.Map) types.List {
+	vals := make(chan types.Value)
+	groupBuilder := types.NewStreamingList(db, vals)
+
+	var group *PhotoGroup
+
+	startGroup := func(first types.Value) {
+		group = &PhotoGroup{
+			Id:     makeID(),
+			Cover:  first,
+			Photos: types.NewSet(),
+		}
+	}
+
+	flush := func() {
+		if group != nil && group.Photos.Len() > 0 {
+			v, err := marshal.Marshal(*group)
+			d.Chk.NoError(err)
+			vals <- v
+			group = nil
+		}
+	}
+
+	lastTime := -math.MaxFloat64
+	byDate.IterAll(func(key, s types.Value) {
+		s.(types.Set).IterAll(func(val types.Value) {
+			dt := float64(key.(types.Number))
+			if (dt - lastTime) > float64(thresh*1e6) {
```

Comment:
Aren't all the dates in this set the same? If so, maybe you could do this check outside the loop for iterating through the set.